clippy_lints/src/hash_borrow_str_semantics.rs

```diff
@@ -0,0 +1,103 @@
+    ///
+    /// Unfortunately it doesn't hold as `hash("abc") != hash("abc".as_bytes())`.
+    ///
+    /// ### Example
+    ///
+    /// ```
+    /// use std::borrow::Borrow;
+    /// use std::hash::{Hash, Hasher};
+    ///
+    /// struct ExampleType {
+    ///     data: String
+    /// }
+    ///
+    /// impl Hash for ExampleType {
+    ///     fn hash<H: Hasher>(&self, state: &mut H) {
+    ///         self.data.hash(state);
+    ///     }
+    /// }
+    ///
+    /// impl Borrow<str> for ExampleType {
+    ///     fn borrow(&self) -> &str {
+    ///         &self.data
+    ///     }
+    /// }
+    ///
+    /// impl Borrow<[u8]> for ExampleType {
+    ///     fn borrow(&self) -> &[u8] {
+    ///         self.data.as_bytes()
+    ///     }
+    /// }
+    /// ```
+    /// As a consequence, hashing a `&ExampleType` and hashing the result of the two
+    /// borrows will result in different values.
+    ///
+    #[clippy::version = "1.75.0"]
+    pub HASH_BORROW_STR_SEMANTICS,
+    correctness,
+    "Ensures that the semantics of Borrow for Hash are satisfied when Borrow<str> and Borrow<[u8]> are implemented"
+}
+
+declare_lint_pass!(HashBorrowStrSemantics => [HASH_BORROW_STR_SEMANTICS]);
+
+impl LateLintPass<'_> for HashBorrowStrSemantics {
+    /// We are emitting this lint at the Hash impl of a type that implements all
+    /// three of Hash, Borrow<str> and Borrow<[u8]>.
+    ///
+    /// We check that we are in the Hash impl
+    /// Then we check that the type implements Borrow<str> and Borrow<[u8]>
+    fn check_item(&mut self, cx: &LateContext<'_>, item: &Item<'_>) {
+        if let ItemKind::Impl(imp) = item.kind
+            && let Some(TraitRef {path: Path {span, res, ..}, ..}) = imp.of_trait
+            && let ty = cx.tcx.type_of(item.owner_id).instantiate_identity()
+            && let Some(hash_id) = cx.tcx.get_diagnostic_item(sym::Hash)
+            && Res::Def(DefKind::Trait, hash_id) == *res
+            && let Some(borrow_id) = cx.tcx.get_diagnostic_item(sym::Borrow)
+            // since we are in the Hash impl, we don't need to check for that.
+            // we need only to check for Borrow<str> and Borrow<[u8]>
+            && implements_trait(cx, ty, borrow_id, &[cx.tcx.types.str_.into()])
+            &&implements_trait(cx, ty, borrow_id, &[Ty::new_slice(cx.tcx, cx.tcx.types.u8).into()])
+        {
+            span_lint_and_then(
+                cx,
+                HASH_BORROW_STR_SEMANTICS,
+                *span,
+                "can't satisfy the semantics of `Hash` when both `Borrow<str>` and `Borrow<[u8]>` are implemented",
+                |diag| {
+                    diag.note("types that implement Hash and Borrow<T> should have equivalent Hash results for borrowed and owned values");
+
+                    diag.note("this is not the case for Borrow<str> and Borrow<[u8]> as the two types result in different Hash values");
```

Comment:
I feel like these two notes could be one. The first one reads a little confusing since it seems unrelated to the situation of having two conflicting `Borrow` impls.
Maybe something along the lines of "`Borrow` requires `Hash` values of all borrowed types to be the same, which is not the case for `Borrow<str>` and `Borrow<[u8]>`?

Reply:
That sounds better, yes!